### PR TITLE
af-packet: add VLAN header when needed in IPS mode

### PR DESCRIPTION
**Pull request by regit**
When packet is coming from a real ethernet card, the kernel is
stripping the vlan header and delivering a modified packet so
we need to insert the VLAN header back before sending the packet
on the wire.

To do so, we pass an option to the raw socket to add a reserve
before the packet data. It will get Suricata some head room to
to move the ethernet addresses before there actual place and
and insert the VLAN header in the correct place.

PR scripts:
* PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/226
* PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/8

Redmine issue https://redmine.openinfosecfoundation.org/issues/1780

### CODE REVIEW

File: src/source-af-packet.c [hunk 1]
```diff
@@ -691,7 +693,21 @@ TmEcode AFPWritePacket(Packet *p)
     if (p->afp_v.peer->flags & AFP_SOCK_PROTECT)
         SCMutexLock(&p->afp_v.peer->sock_protect);
     socket = SC_ATOMIC_GET(p->afp_v.peer->socket);
-    if (sendto(socket, GET_PKT_DATA(p), GET_PKT_LEN(p), 0,
+    /* We have a vlan ID without header so first VLAN have been stripped */
+    if (p->vlan_idx && !p->vlanh[0] && p->afp_v.relptr) {
+        pstart = GET_PKT_DATA(p) - VLAN_HEADER_LEN;
+        plen = GET_PKT_LEN(p) + VLAN_HEADER_LEN;
+        /* move ethernet addresses */
+        memmove(pstart, GET_PKT_DATA(p), 2 * ETH_ALEN);
```
Comment (inliniac) [inline src/source-af-packet.c]: Does this code get called when we are not in zero copy mode? In that case the reserve call won't do much for us, will it? I worry that GET_PKT_DATA will then point to a suricata-allocated memory block and memmove will write before it.

Comment (regit) [inline src/source-af-packet.c]: The test is ` if (p->vlan_idx && !p->vlanh[0] && p->afp_v.relptr)` and the last test in there check that we have the release function set. And we have this release pointer set only if we are in zero copy mode.

**Comment by inliniac** [comment]
If no one tests this in time for 3.2.1 we'll merge it in the dev branch and backport later if no bug reports.

**Comment by gozzy** [comment]
I hope that we'll try it out today.

**Comment by gozzy** [comment]
The fix itself is fine, VLAN tags are correctly passed when working in inline mode.

However, we've noticed another problem: VLAN-tagged packets seem to loose 'priority' and 'CFI' attributes (https://wiki.wireshark.org/VLAN), probably because of this:

p->vlan_id[0] = h.h2->tp_vlan_tci & 0x0fff;

The mask 0x0fff cuts those attributes and reconstructed packet slightly differs from the original one. I don't know whether it's critical, but probably this should be kept in mind for a later fix.

**Comment by inliniac** [comment]
On 25-01-17 14:27, Alexander Gozman wrote:
> The fix itself is fine, VLAN tags are correctly passed when working in
> inline mode.
> 
> However, we've noticed another problem: VLAN-tagged packets seem to
> loose 'priority' and 'CFI' attributes (https://wiki.wireshark.org/VLAN),
> probably because of this:
> 
> p->vlan_id[0] = h.h2->tp_vlan_tci & 0x0fff;
> 
> The mask 0x0fff cuts those attributes and reconstructed packet slightly
> differs from the original one. I don't know whether it's critical, but
> probably this should be kept in mind for a later fix.
> 

Great catch, thanks Alex. I'd like to see this fixed before merge.

**Comment by gozzy** [comment]
I can try to make another PR, but I have some doubts about a proper way to handle this case. The most obvious idea is to add two bitfields into the Packet structure and store the attributes there. Like:

struct Packet {
    ...
    uint16_t vlan_id[2];
    uint8_t vlan_idx;
    ....
}

Then, in source-af-packet.c (and, maybe, source-pfring.c):

...
p->priority = (h.h2->tp_vlan_tci & 0xe000) >> 13;
p->cfi = (h.h2->tp_vlan_tci & 0x1000) >> 12;
...

Finally, when reconstructing a packet:

uint16_t tag = htons(p->vlan_id[0] | (p->priority << 13) | (p->cfi << 12));
*(uint16_t *)(pstart + 2 * ETH_ALEN + 2) = tag;

**Comment by gozzy** [comment]
Or, maybe, we should just remove the mask 0x0fff and store a whole tag. Anyway, if one needs to get VLAN ID, priority or CFI, suricata provides macros like GET_VLAN_ID, GET_VLAN_PRIORITY or GET_VLAN_CFI.

**Comment by inliniac** [comment]
I don't think the rest of the code is expecting the prio & cfi to be part of the vlan_id field. I think I prefer to add 2 new fields like you suggested in https://github.com/inliniac/suricata/pull/2405#issuecomment-275146178

Maybe we can store the vlan_id as vlan_id:12, although I'm not sure how that works in an array.

**Comment by regit** [comment]
Good catch indeed @gozzy. Do you want me to take code update ?

**Comment by gozzy** [comment]
@regit, it would be great if you could make it 'cause at the moment I don't realise all the details about VLAN handling. It's done in source-af-packet/pfring, in DecodeVLAN function...

**Comment by inliniac** [comment]
Needs improvements

**Comment by inliniac** [comment]
I think the priority and cfi could be saved in Packet::AFPPacketVars. This issue is afpacket specific so we can use a afpacket specific store. Think that would make a fix much easier. Care to try @gozzy ?

**Comment by gozzy** [comment]
Well, I can try if @regit is not already updating the patch.

**Comment by inliniac** [comment]
@regit seems rather busy and since you have a test setup I think it may be easiest if you try, but I'll let @regit speak for himself :)

Thanks for considering anyway!

**Comment by regit** [comment]
OK, I'm getting a look at it right now. But @gozzy, you will be the one to test it if ok with you.

**Pull request closed without merging**